File: desktop-viewer/simpleperf-viewer/device-adb/src/main/kotlin/com/androidperformancestudio/adb/AdbLocator.kt
package com.androidperformancestudio.adb

import com.androidperformancestudio.model.ErrorCategory
import com.androidperformancestudio.model.StudioError
import com.androidperformancestudio.model.StudioResult
import com.androidperformancestudio.toolchain.HostOperatingSystem
import com.androidperformancestudio.toolchain.HostPlatform
import java.io.File
import java.nio.file.Files
import java.nio.file.Path

data class AdbConfiguration(
    val userConfiguredPath: Path? = null,
    val androidSdkPath: Path? = null,
)

enum class AdbLocationSource {
    USER_CONFIGURATION,
    ANDROID_HOME,
    ANDROID_SDK_ROOT,
    PATH,
    DEFAULT_SDK,
}

data class AdbLocation(
    val executable: Path,
    val source: AdbLocationSource,
)

class SystemAdbLocator(
    private val platform: HostPlatform,
    private val environment: Map<String, String> = System.getenv(),
    private val userHome: Path = Path.of(System.getProperty("user.home")),
    private val pathSeparator: String = File.pathSeparator,
    private val isUsableExecutable: (Path) -> Boolean = { path ->
        Files.isRegularFile(path) &&
            (platform.operatingSystem == HostOperatingSystem.WINDOWS || Files.isExecutable(path))
    },
) {
    fun locate(configuration: AdbConfiguration = AdbConfiguration()): StudioResult<AdbLocation> =
        configuration.userConfiguredPath?.let(::locateConfigured)
            ?: configuration.androidSdkPath?.let(::locateConfiguredSdk)
            ?: locateAutomatically()

    private fun locateConfigured(configured: Path): StudioResult<AdbLocation> =
        if (isUsableExecutable(configured)) {
            success(configured, AdbLocationSource.USER_CONFIGURATION)
        } else {
            failure(
                code = "ADB_CONFIGURED_PATH_INVALID",
                message = "Configured adb executable is not usable: $configured",
            )
        }

    private fun locateConfiguredSdk(sdkDirectory: Path): StudioResult<AdbLocation> {
        val adb = sdkDirectory.resolve("platform-tools").resolve(executableName)
        return if (isUsableExecutable(adb)) {
            success(adb, AdbLocationSource.USER_CONFIGURATION)
        } else {
            failure(
                code = "ADB_CONFIGURED_SDK_INVALID",
                message = "Configured Android SDK doesn't contain a usable platform-tools/$executableName: $sdkDirectory",
            )
        }
    }

    private fun locateAutomatically(): StudioResult<AdbLocation> {
        val candidate = candidates().firstOrNull { isUsableExecutable(it.path) }
        return candidate?.let { success(it.path, it.source) }
            ?: failure(
                code = "ADB_NOT_FOUND",
                message = "adb wasn't found in Android SDK, PATH, or the default SDK directory",
            )
    }

    private fun candidates(): Sequence<AdbCandidate> =
        sequence {
            sdkCandidate("ANDROID_HOME", AdbLocationSource.ANDROID_HOME)?.let { yield(it) }
            sdkCandidate("ANDROID_SDK_ROOT", AdbLocationSource.ANDROID_SDK_ROOT)?.let { yield(it) }
            pathCandidates().forEach { yield(it) }
            defaultSdkDirectories().forEach { sdkDirectory ->
                yield(
                    AdbCandidate(
                        sdkDirectory.resolve("platform-tools").resolve(executableName),
                        AdbLocationSource.DEFAULT_SDK,
                    ),
                )
            }
        }

    private fun sdkCandidate(
        variable: String,
        source: AdbLocationSource,
    ): AdbCandidate? =
        environmentValue(variable)
            ?.let(::environmentPath)
            ?.resolve("platform-tools")
            ?.resolve(executableName)
            ?.let { AdbCandidate(it, source) }

    private fun pathCandidates(): Sequence<AdbCandidate> =
        environmentValue("PATH")
            .orEmpty()
            .splitToSequence(pathSeparator)
            .filter(String::isNotBlank)
            .mapNotNull(::environmentPath)
            .map { directory -> AdbCandidate(directory.resolve(executableName), AdbLocationSource.PATH) }

    private fun defaultSdkDirectories(): List<Path> =
        when (platform.operatingSystem) {
            HostOperatingSystem.MACOS -> listOf(userHome.resolve("Library/Android/sdk"))
            HostOperatingSystem.LINUX -> listOf(userHome.resolve("Android/Sdk"))
            HostOperatingSystem.WINDOWS ->
                listOf(
                    environmentValue("LOCALAPPDATA")
                        ?.let(::environmentPath)
                        ?.resolve("Android/Sdk")
                        ?: userHome.resolve("AppData/Local/Android/Sdk"),
                )
        }

    private fun environmentValue(name: String): String? =
        if (platform.operatingSystem == HostOperatingSystem.WINDOWS) {
            environment.entries.firstOrNull { (key, _) -> key.equals(name, ignoreCase = true) }?.value
        } else {
            environment[name]
        }?.takeIf(String::isNotBlank)

    private fun environmentPath(rawPath: String): Path? {
        val unquoted = rawPath.trim().removeSurrounding("\"")
        val expanded =
            if (platform.operatingSystem == HostOperatingSystem.WINDOWS) {
                WINDOWS_ENVIRONMENT_VARIABLE.replace(unquoted) { match ->
                    environmentValue(match.groupValues[1]) ?: match.value
                }
            } else {
                unquoted
            }
        return runCatching { Path.of(expanded) }.getOrNull()
    }

    private val executableName: String
        get() = if (platform.operatingSystem == HostOperatingSystem.WINDOWS) "adb.exe" else "adb"

    private fun success(
        path: Path,
        source: AdbLocationSource,
    ): StudioResult.Success<AdbLocation> =
        StudioResult.Success(
            AdbLocation(
                executable = path.toAbsolutePath().normalize(),
                source = source,
            ),
        )

    private fun failure(
        code: String,
        message: String,
    ): StudioResult.Failure =
        StudioResult.Failure(
            StudioError(
                category = ErrorCategory.CONFIGURATION,
                code = code,
                message = message,
            ),
        )

    private data class AdbCandidate(
        val path: Path,
        val source: AdbLocationSource,
    )

    private companion object {
        val WINDOWS_ENVIRONMENT_VARIABLE = Regex("%([^%]+)%")
    }
}

File: electron-viewer/packages/platform-adb/src/adb-locator.ts
import { statSync } from 'node:fs';
import { homedir } from 'node:os';
import { delimiter, join, normalize } from 'node:path';
import { AdbNotFoundError, AdbNotExecutableError } from './adb-errors.js';

export interface AdbLocatorConfiguration {
  readonly executablePath?: string;
  readonly androidSdkPath?: string;
}

export type AdbLocationSource =
  | 'EXPLICIT_EXECUTABLE'
  | 'EXPLICIT_SDK'
  | 'ADB_ENVIRONMENT'
  | 'ANDROID_HOME'
  | 'ANDROID_SDK_ROOT'
  | 'PATH'
  | 'DEFAULT_SDK';

export interface AdbLocation {
  readonly executable: string;
  readonly source: AdbLocationSource;
}

export interface AdbLocatorDependencies {
  readonly env: Readonly<Record<string, string | undefined>>;
  readonly userHome: string;
  readonly platform: NodeJS.Platform;
  readonly pathSeparator: string;
  readonly isUsableExecutable: (path: string) => boolean;
  readonly join: (...parts: string[]) => string;
  readonly normalize: (path: string) => string;
}

function defaultIsUsableExecutable(path: string): boolean {
  try {
    const stats = statSync(path);
    if (!stats.isFile()) return false;
    return process.platform === 'win32' || (stats.mode & 0o111) !== 0;
  } catch {
    return false;
  }
}

export function defaultAdbLocatorDependencies(): AdbLocatorDependencies {
  return {
    env: process.env,
    userHome: homedir(),
    platform: process.platform,
    pathSeparator: delimiter,
    isUsableExecutable: defaultIsUsableExecutable,
    join,
    normalize,
  };
}

function isWindows(platform: NodeJS.Platform): boolean {
  return platform === 'win32';
}

function isMac(platform: NodeJS.Platform): boolean {
  return platform === 'darwin';
}

export class AdbExecutableLocator {
  private readonly deps: AdbLocatorDependencies;

  constructor(dependencies: AdbLocatorDependencies = defaultAdbLocatorDependencies()) {
    this.deps = dependencies;
  }

  locate(configuration: AdbLocatorConfiguration = {}): AdbLocation {
    if (configuration.executablePath !== undefined) {
      return this.locateConfiguredExecutable(configuration.executablePath);
    }
    if (configuration.androidSdkPath !== undefined) {
      return this.locateConfiguredSdk(configuration.androidSdkPath);
    }
    for (const candidate of this.candidates()) {
      if (this.deps.isUsableExecutable(candidate.executable)) return candidate;
    }
    throw new AdbNotFoundError();
  }

  private get executableName(): string {
    return isWindows(this.deps.platform) ? 'adb.exe' : 'adb';
  }

  private locateConfiguredExecutable(executable: string): AdbLocation {
    if (!this.deps.isUsableExecutable(executable)) throw new AdbNotExecutableError(executable);
    return { executable, source: 'EXPLICIT_EXECUTABLE' };
  }

  private locateConfiguredSdk(sdk: string): AdbLocation {
    const executable = this.deps.join(sdk, 'platform-tools', this.executableName);
    if (!this.deps.isUsableExecutable(executable)) throw new AdbNotExecutableError(executable);
    return { executable, source: 'EXPLICIT_SDK' };
  }

  private environmentValue(name: string): string | undefined {
    if (isWindows(this.deps.platform)) {
      const entry = Object.entries(this.deps.env).find(([key]) => key.toLowerCase() === name.toLowerCase());
      return entry?.[1]?.trim().length ? entry[1] : undefined;
    }
    const value = this.deps.env[name];
    return value !== undefined && value.trim().length > 0 ? value : undefined;
  }

  private expandPath(value: string): string {
    const unquoted = value.trim().replace(/^"(.*)"$/, '$1');
    if (!isWindows(this.deps.platform)) return unquoted;
    return unquoted.replace(/%([^%]+)%/g, (match, name: string) => this.environmentValue(name) ?? match);
  }

  private sdkCandidate(name: string, source: AdbLocationSource): AdbLocation | undefined {
    const root = this.environmentValue(name);
    if (root === undefined) return undefined;
    return {
      executable: this.deps.join(this.expandPath(root), 'platform-tools', this.executableName),
      source,
    };
  }

  private defaultSdkDirectories(): string[] {
    if (isWindows(this.deps.platform)) {
      const localAppData = this.environmentValue('LOCALAPPDATA');
      return [localAppData !== undefined ? this.deps.join(localAppData, 'Android/Sdk') : this.deps.join(this.deps.userHome, 'AppData/Local/Android/Sdk')];
    }
    if (isMac(this.deps.platform)) {
      return [this.deps.join(this.deps.userHome, 'Library/Android/sdk')];
    }
    return [this.deps.join(this.deps.userHome, 'Android/Sdk')];
  }

  private candidates(): AdbLocation[] {
    const seen = new Set<string>();
    const result: AdbLocation[] = [];
    const emit = (location: AdbLocation | undefined): void => {
      if (location === undefined) return;
      const key = this.deps.normalize(location.executable);
      if (seen.has(key)) return;
      seen.add(key);
      result.push(location);
    };
    const adbEnv = this.environmentValue('ADB');
    if (adbEnv !== undefined) emit({ executable: this.expandPath(adbEnv), source: 'ADB_ENVIRONMENT' });
    const adbPath = this.environmentValue('ADB_PATH');
    if (adbPath !== undefined) emit({ executable: this.expandPath(adbPath), source: 'ADB_ENVIRONMENT' });
    emit(this.sdkCandidate('ANDROID_HOME', 'ANDROID_HOME'));
    emit(this.sdkCandidate('ANDROID_SDK_ROOT', 'ANDROID_SDK_ROOT'));
    const pathValue = this.environmentValue('PATH');
    if (pathValue !== undefined) {
      for (const entry of pathValue.split(this.deps.pathSeparator)) {
        if (entry.trim().length === 0) continue;
        emit({ executable: this.deps.join(this.expandPath(entry), this.executableName), source: 'PATH' });
      }
    }
    for (const directory of this.defaultSdkDirectories()) {
      emit({
        executable: this.deps.join(directory, 'platform-tools', this.executableName),
        source: 'DEFAULT_SDK',
      });
    }
    return result;
  }
}
